### PR TITLE
Simplifies UserAccountManager's implementation of auth

### PR DESCRIPTION
**Pull request by trooper2013**
This PR is part of the overall theme for simplification of auth flows. Previous PR simplified endpoint token calls. This one simplifies the login flow(basic & adv. auth).

1. Removes usage of SFSDKOAuthClient with the SFUserAccountManager. The SFUserAccountManager now interacts directly with SFOAuthCoordinator.  The state is captured as an ongoing authentication session. 

2. @wmathurin FYI:  I could not leverage the SFSDKAppConfig as discussed due to its design. So internally I continue to use SFSDKAuthRequest to kick off an authentication request. We can perhaps improve it in the near future.  The design keeps the coordinators from assumptions  regarding single/multiple auth sessions and also the overall request ( loginhost, client id...) .  Both the SFSDKAuthRequest & SFSDKAuthSession are internal.

3. Retracted the SFSDKAuthErrorManager. This is not really used externally and should not have been public. Refactored the error processing to use Auth Session.

4. I have tested basic auth & adv. auth with all combinations of user switching/host changes,  passcode flow and navigation. Ive even tried a few error conditions. 

NOTE: IDP flow refactoring is pending  (Next PR), This is when i will physically remove the non-required classes.

### CODE REVIEW

**Comment by bhariharan** [comment]
@trooper2013 I think there are conflicts in a file. You will need to merge in @bbirman's changes.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFOAuthCoordinator+Internal.h [hunk 1]
```diff
@@ -23,32 +23,37 @@
...
     SFOAuthTokenEndpointFlowAdvancedBrowser
 };
+NS_ASSUME_NONNULL_BEGIN
 
 @interface SFOAuthCoordinator ()
 
 @property (assign) BOOL authenticating;
-@property (nonatomic, strong, readonly) NSURLSession *session;
-@property (nonatomic, strong) NSMutableData *responseData;
+@property (nonatomic, strong, readonly, nullable) NSURLSession *session;
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFOAuthCoordinator+Internal.h]: Should these be `nullable`? I feel like at least some of these should be `nonnull`, like `NSURLSession` and `WKWebView`, for instance. One can't authenticate or perform any actions without either of those.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFSDKAuthRequest.m [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+#import "SFSDKAuthRequest.h"
+#import "SFSDKLoginViewControllerConfig.h"
+#import "SFSDKAppLockViewConfig.h"
+@implementation SFSDKAuthRequest
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFSDKAuthRequest.m]: No implementation?

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccountManager.m [hunk 3]
```diff
@@ -89,12 +93,17 @@
 
 NSString * const kSFSDKUserAccountManagerErrorDomain = @"com.salesforce.mobilesdk.SFUserAccountManager";
 
+static NSString * const kSFInvalidCredentialsAuthErrorHandler = @"InvalidCredentialsErrorHandler";
+static NSString * const kSFConnectedAppVersionAuthErrorHandler = @"ConnectedAppVersionErrorHandler";
+static NSString * const kSFNetworkFailureAuthErrorHandler = @"NetworkFailureErrorHandler";
+static NSString * const kSFGenericFailureAuthErrorHandler = @"GenericFailureErrorHandler";
+static NSString * const kSFRevokePath = @"/services/oauth2/revoke";
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccountManager.m]: I thought we captured all endpoints in the `OAuthClient` class that we recently created. Shouldn't revocation be a part of that class too? I don't think this should live here.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccountManager.m [hunk 4]
```diff
@@ -371,6 +434,32 @@ - (void)logoutUser:(SFUserAccount *)user {
     }];
 }
 
+- (void)revokeRefreshToken:(SFOAuthCredentials *)credentials {
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccountManager.m]: This should probably move to the `OAuthClient` class - it's related to credentials, not really associated with the concept of a logged in user.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccountManager.m [hunk 5]
```diff
@@ -465,248 +552,239 @@ + (BOOL)errorIsInvalidAuthCredentials:(NSError *)error {
...
-    //configuration was realized from org/domain setting.
-    if (!client.config.useBrowserAuth) {
-        NSString *newKey = [SFSDKOAuthClientCache keyFromCredentials:client.credentials type:SFOAuthClientKeyTypeAdvanced];
-        [[SFSDKOAuthClientCache sharedInstance] removeClient:client];
-        [[SFSDKOAuthClientCache sharedInstance] addClient:client forKey:newKey];
+    SFLoginViewController *loginViewController = [self createLoginViewControllerInstance:coordinator];
+    loginViewController.oauthView = view;
+    SFSDKAuthViewHolder *viewHolder = [SFSDKAuthViewHolder new];
+    viewHolder.loginController = loginViewController;
+    //self.authContext.viewHolder = viewHolder;
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccountManager.m]: Remove commented out code.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFSDKAuthRequest.m (hunk 2 above)
Comment (trooper2013) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFSDKAuthRequest.m]: @bhariharan none required. Just a set of properties.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFOAuthCoordinator+Internal.h (hunk 1 above)
Comment (trooper2013) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFOAuthCoordinator+Internal.h]: Agree, these were always nullable. In introduced the NS_ASSUME_NONNULL to suppress warnings. These are internal properties. Session is also nullified/cleared at some point.

File: libs/SalesforceSDKCore/SalesforceSDKCore.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -118,8 +118,14 @@
...
+		B72171522353BFD90022510F /* SFSDKAuthSession.h in Headers */ = {isa = PBXBuildFile; fileRef = B72171502353BFD90022510F /* SFSDKAuthSession.h */; settings = {ATTRIBUTES = (Private, ); }; };
+		B72171532353BFD90022510F /* SFSDKAuthSession.m in Sources */ = {isa = PBXBuildFile; fileRef = B72171512353BFD90022510F /* SFSDKAuthSession.m */; };
+		B72171562353BFF20022510F /* SFSDKAuthRequest.h in Headers */ = {isa = PBXBuildFile; fileRef = B72171542353BFF20022510F /* SFSDKAuthRequest.h */; settings = {ATTRIBUTES = (Private, ); }; };
+		B72171572353BFF20022510F /* SFSDKAuthRequest.m in Sources */ = {isa = PBXBuildFile; fileRef = B72171552353BFF20022510F /* SFSDKAuthRequest.m */; };
 		B722A709233C437E0089736E /* RestClient.swift in Sources */ = {isa = PBXBuildFile; fileRef = B722A708233C437E0089736E /* RestClient.swift */; };
 		B7352CA522761D8400DA2CFF /* SFManagedPreferencesTest.m in Sources */ = {isa = PBXBuildFile; fileRef = B7352CA422761D8400DA2CFF /* SFManagedPreferencesTest.m */; };
+		B736CB57234D50FE00337189 /* SFSDKAuthUI.h in Headers */ = {isa = PBXBuildFile; fileRef = B736CB55234D50FE00337189 /* SFSDKAuthUI.h */; };
```
Comment (bbirman) [inline libs/SalesforceSDKCore/SalesforceSDKCore.xcodeproj/project.pbxproj]: What's SFSDKAuthUI?

Comment (trooper2013) [reply libs/SalesforceSDKCore/SalesforceSDKCore.xcodeproj/project.pbxproj]: wip. Thanks for catching it.

**Pull request merged**